### ISSUE TITLE
Incorrect date information

### ISSUE THREAD

**Issue by iago1460** [django-recurrence/django-recurrence#70]
I have notice that the widget send Rdates, Exdates and Until dates in UTC format to the server like '20160725T230000Z' but looks like the widget got the timezone info from the browser instead of from the [timezone setting](https://docs.djangoproject.com/en/1.9/ref/settings/#std:setting-TIME_ZONE)

Other issue is that when I access the field information the dates don't have a timezone and it looks like they were converted to the current one.

My browser timezone is UTC+1 and my django TIME_ZONE UTC+2
So in my case the widget send '20160725T230000Z' (datetime.datetime(2016, 7, 25, 23, 0, tzinfo=tzutc())) and when I access the field it contains datetime.datetime(2016, 7, 26, 1, 0).

Any ideas, can this be fixed somehow? Thanks

**Comment by audiolion** [comment]
I have the same issue, the widget does indeed use the browsers timezone setting, but serializes with the django timezone setting (by using pytz).

**Comment by csetarun** [comment]
I am also having the same issue.. the value which is saved is one day less than the required.

**Comment by audiolion** [comment]
I actually solved this by including settings.TIMEZONE = 'My Tz', i didnt realize it wasnt in settings so dates werent converted back properly

**Comment by KonstantinSchubert** [comment]
I think that this PR https://github.com/django-recurrence/django-recurrence/pull/104 fixes the issue.

@audiolion 
> but serializes with the django timezone setting (by using pytz).
Are you sure about this? As far as I can see, it just strips the timezone information when serializing.

**Comment by iago1460** [comment]
My current solution was to remove the timezone information from the frontend and letting the backend to convert that date to the current timezone.

[diff link](https://github.com/django-recurrence/django-recurrence/compare/master...iago1460:master]) (ignore the changes on recurrence-widget.js)

**Comment by GriceTurrble** [comment]
Coming here with the same issue.

More generally, why is this package serializing time information in an RDATE in the first place, even while the frontend widget does not include a time selector?

My team is using this package in an application where we need a set of dates, nothing more. We handle time and timezone separately so that a user, for instance, in Asia can write details relevant to an object located in the EU (saving *that* timezone), while this data is stored on a server in the US. They select the timezone for this object, they select the time of day relevant to *that* timezone, and they select the dates; while we do the math and combine all three in the backend.

Meanwhile, this package takes in RDATEs as datetimes set to midnight in the user's *browser's* timezone, converts that to UTC (on the frontend), then serializes it in the database, then reads it back and automatically converts it to the *server local timezone* and presents those RDATEs as naive datetimes. What in tarnation?

If you're going to serialize time data, please add a time selector with timezone selector to perform the math properly, and/or include additional parameters from the RFC 2445 spec to indicate the timezone used for a datetime without converting it to UTC. Either that, or drop time serialization entirely and just use the bare date: `RDATE:20180613` ought to be valid and produce a simpler `datetime.date` object.

**Comment by dominicrodger** [comment]
@GriceTurrble - nice to hear you're using django-recurrence, thanks for the feedback!

If you want django-recurrence to serialize time data differently, please send a pull request. This particular behaviour hasn't ever bitten me (or presumably, any of the other maintainers), so we've never addressed it. As far as I know, we're all working on this in our spare time, by the sounds of your comment you're using it at work*, so perhaps you might have time to work on finding a way forwards here?

* apologies if I misinterpreted your comment :)

**Comment by GriceTurrble** [comment]
Yes, this is being used in a work context, though to contribute I'd have to do so on personal time.

That said, I noticed that we are actually using an older version of django-recurrence (1.2.0). Please take what I said earlier with a big grain of salt, as I may be way off base here. Sorry!

**Comment by GriceTurrble** [comment]
Off-topic, but my issue is less about time and more about code ownership. Best to keep things separated to avoid potential problems down the line.

On-topic: I've upgraded to the latest and the same issue exists, with an RDATE being serialized with time information using the user's local machine settings, converted to UTC, and all the rest.

I propose, and aim to work on:

* Change the frontend widget/s so that it/they serialize dates as bare dates, without adding time data of any kind. Unless someone else wants to add a widget that collects time data, which I don't think is necessary, then any JS code calculating time can be stripped down.
* Allow backend to store `datetime.date` objects in the form `'RDATE:YYYYMMDD'`, and de-serializing this format back to `datetime.date` objects.
* If a naive datetime is serialized, it should be localized on server local timezone settings if present (UTC if not). If it is aware already, pass.
* Serialize datetime using the second format described in [RFC 2445](https://www.ietf.org/rfc/rfc2445.txt) (Section "4.8.5.3"), which uses the `TZID` property to identify the timezone by name.
  * Continue to support reading the first format, which is already in use, for backwards compatibility.
* De-serialize datetimes as timezone-aware by attaching the same timezone originally serialized (perform no implicit conversion to server local timezone)
  * Optionally add a property on the recurrence field for "utc_rdates" that presents the list of rdates converted to UTC (as a read-only property).

All that and hopefully nothing blows up. :)

**Comment by dominicrodger** [comment]
Thanks @GriceTurrble. That all sounds good to me. We may want to make this behaviour opt in for at least one release. One thing I'm a little confused about - if we're stripping out times entirely (per your first point), under what circumstances would we be serializing datetimes instead of dates?

**Comment by GriceTurrble** [comment]
Without adding a time selector to the widget, the benefit would be for the
backend. Users may choose to create recurrences programmatically by storing
datetime objects, which can already be serialized with time data. It
wouldn't make sense to remove this functionality as an option.

**Comment by pasevin** [comment]
Hi guys, has this been considered for the near future? Or is this project not maintained anymore? Trying to figure out if it's worth adopting it. Looks like the timezone issue is a big one.

**Comment by dominicrodger** [comment]
This isn't something I'm looking at @pasevin - it's never caused me any issues personally, and I couldn't reproduce the issue when I tried IIRC.

As to whether the library is maintained - it has quite a few active users I think, but I'm no longer doing any feature development or bug fixes myself, as I no longer use this library. That said, I do my best to merge pull requests in a timely fashion, and respond to issues when I see them.

**Comment by pasevin** [comment]
> My current solution was to remove the timezone information from the frontend and letting the backend to convert that date to the current timezone.
> 
> [diff link](https://github.com/django-recurrence/django-recurrence/compare/master...iago1460:master) (ignore the changes on recurrence-widget.js)

I tried it out at it fixes this annoying issue! I think this would be a great addition. I will create a pull request based on this and let the @dominicrodger decide.
